### DISCUSSION TITLE
WASM/WASI Support

### DISCUSSION THREAD

**Discussion by cwarden** [charmbracelet/bubbletea#300] [Ideas]
[_Migrated from [Charm Slack Workspace](https://charmbracelet.slack.com/archives/CQ2E9CGTV/p1649636731692529)_]

It would be nice to be able to run bubbletea applications in a browser.

[tinygo](https://tinygo.org/) supports targeting [WASI](https://wasi.dev/), but there are some terminal dependencies that don't work.

[hackpad.org](http://hackpad.org/) is a slightly different approach, but trying to build a simple application, e.g. https://github.com/charmbracelet/bubbletea/blob/master/examples/simple/main.go, illustrates some of the same issues that need to be addressed.

If anyone wants to work on this, I can contribute toward a bounty.

**Comment by cwarden** [comment]
In meantime, I've found another solution for running a bubbletea application in a browser by compiling it using i386 arch and running it in v86.  Here's an example: https://kalibrate.app/
It should be possible to slim down the image to reduce the time to start.

**Comment by BigJk** [comment]
@cwarden I build a POC of which changes would be necessary to get Bubbletea working in the browser via WASM and the stock golang compiler

https://github.com/BigJk/bubbletea-in-wasm

**Reply by cwarden** [reply]
Kudos.  This is fantastic!

**Comment by joeblew999** [comment]
Related to getting WASM working officially : https://github.com/charmbracelet/bubbletea/issues/1410

a web version that runs as WASM.

It’s  also possible to take the dsl of charm and generate pure html at runtime , using charm running as a service worker or on a server to server the dsl . This would require more work and possibly a go generate phase for bootstrap of the web things out of the DSl into concrete golang files and etcs needed . Then another go build / run on top .

Htmx pattern is the way the web and backend layer run as. REPL, so that the gui can be changed at runtime just like a htmx web project .

**Reply by cwarden** [reply]
Another option is to use bubbletea's implementation of the Elm architecture, but have it render html rather than text.  This is the approach I've been working on with [masc](https://github.com/octoberswimmer/masc).

**Reply by joeblew999** [reply]
like the way you interact with it in the todo:

```go

package components

import (
	"github.com/octoberswimmer/masc"
	"github.com/octoberswimmer/masc/elem"
	"github.com/octoberswimmer/masc/event"
	"github.com/octoberswimmer/masc/prop"
)
```
